lib/Driver/Driver.cpp

```diff
@@ -268,6 +268,47 @@ class Driver::InputInfoMap
 };
 using InputInfoMap = Driver::InputInfoMap;
 
+// Get the filename for build record.
+// Additionally set 'outputBuildRecordForModuleOnlyBuild' to true if this is
+// full compilation with swiftmodule.
+static bool getCompilationRecordPath(std::string &buildRecordPath,
+                                     bool &outputBuildRecordForModuleOnlyBuild,
+                                     const OutputInfo &OI,
+                                     const Optional<OutputFileMap> &OFM,
+                                     DiagnosticEngine *Diags) {
+  if (!OFM) {
+    // FIXME: This should work without an output file map. We should have
+    // another way to specify a build record and where to put intermediates.
+    if (Diags)
+      Diags->diagnose(SourceLoc(), diag::incremental_requires_output_file_map);
+    return true;
+  }
+
+  if (auto *masterOutputMap = OFM->getOutputMapForSingleOutput())
+    buildRecordPath = masterOutputMap->lookup(file_types::TY_SwiftDeps);
+
+  if (buildRecordPath.empty()) {
+    if (Diags)
+      Diags->diagnose(SourceLoc(),
+                      diag::incremental_requires_build_record_entry,
+                      file_types::getTypeName(file_types::TY_SwiftDeps));
+    return true;
+  }
+
+  // In 'emit-module' only mode, use build-record filename suffixed with
+  // '~moduleonly'. So that module-only mode doesn't mess up build-record
+  // file for full compilation.
+  if (OI.CompilerOutputType == file_types::TY_SwiftModuleFile) {
+    buildRecordPath = buildRecordPath.append("~moduleonly");
+  } else if (OI.ShouldTreatModuleAsTopLevelOutput) {
+    // If we emit module along with full compilation, emit build record
+    // file for '-emit-module' only mode as well.
+    outputBuildRecordForModuleOnlyBuild = true;
+  }
+
+  return false;
+}
```

Comment:
I like this refactoring! This refactoring has opened my eyes to a further refactoring! It looks like if the routine returns false (no error) `buildRecordPath` will not be empty. It also looks like the condition for `outputBuildRecordForModuleOnlyBuild` is an invariant function on OI.
So what this suggests is that the routine not deal with ``outputBuildRecordForModuleOnlyBuild``. Instead create a function to compute that predicate. That function would be a member function of the class of OI (OutputInfo?).
Just call that function where you need the value. It is cheap enough.
Then, you can simplify this routine to just return the `buildRecordPath` with an empty string returned value signaling error. Cool!!

Reply:
Thanks! That's sounds good.
I'll do it in a follow-up.